### ISSUE TITLE
Incompatibility with Android ST client

### ISSUE THREAD

**Issue by KurtSanders** [KurtSanders/STAmbientWeather#3]
The Ambient Weather Tile DTH uses an extensive preference section to configure required features and API keys.  Android devices running the classic ST mobile client do not appear to support a DTH preference section as I have developed it, where as iOS devices do work as expected.  

I intend to re-write this Ambient Weather Tile into a SmartApp **and** a DTH to provide android compatibility, although I do not have such a device from which to test.

**Comment by KurtSanders** [comment]
A **V2 version of Ambient Weather Reporter V2** is now available for immediate testing on Android devices.  I have tested on iOS phone and it works.  

The V2 version requires **both** a _SmartApp Service Manager_ and the new _V2 Ambient Weather Station Device Handler_.  The updated readme in the BETA branch is current and it would be appreciated for anyone to test to see if it works and displays on Android ST client.

The V2 can be installed along side of your V1 tile, so do NOT delete your working V1 device handler until we get this fully tested.

Here is the [link to the beta branch](https://github.com/KurtSanders/STAmbientWeather/tree/beta) and detailed instructions.

https://github.com/KurtSanders/STAmbientWeather/tree/beta

**Comment by jeffp1105** [comment]
Hi Kurt,
Thanks for creating the new version. I've been working with it and have some following observations you may find useful. BTW, I am on a Samsung S9/Android V8.0(Oreo).

- I had some minor install problems. I used the desktop IDE to set the API and APP keys. When installing the app on the phone, I got the Success message but when I hit "Next" I got an error saying I had to set the other required fields. However, the page for entering those values did not appear. TEMP WORKAROUND: by changing lines 80, 84, 90 to not required. I could get to the next page.

- I get occasional timeouts but I think that is on the server side as I have the update at 1 minutes

-  The good: The logs all look good, I can see the data, the api success, etc, etc. other than the warning from above.  Also, on the "My Things" page, I see the station at the top with the correct temperature to the 0.1 degree. I can add the tile to My Dashboard and see the top.

- The bad:  When I tap the tile in the dashboard or the Thing in my thing list, I get an error
![screenshot_20181129-160121_smartthings](https://user-images.githubusercontent.com/44684349/49259152-6c0b0a00-f3f5-11e8-9290-ff5de20ee75f.jpg)
. The logs don't seem to provide any useful data.

Let me know what I can do to help.
Thanks again
Jeff

**Comment by jeffp1105** [comment]
Oh, but the Recently list shows all the data. See attached.
![screenshot_20181129-164110_smartthings](https://user-images.githubusercontent.com/44684349/49259217-bab8a400-f3f5-11e8-9e7e-29a08861d544.jpg)

**Comment by KurtSanders** [comment]
Dear Jeff,

Thanks for assisting with this debugging, your information was very helpful.  

I have created a **new DTH** named **"Ambient Weather Station - Android"** to hopefully work with the V2 SmartApp.  I removed what I think are the issues, complex color and icon subroutines, that cause the Android client to error.   

I also made the changes in the **Ambient Weather Station Service Manager SmartApp** from 'true' to 'false' as you suggested so you can test.   Please do the following:

1. Perform an **Update from Repository** in the ST IDE to the following:
- [ ] Ambient Weather Station Service Manager _(Update in My SmartApps)_
- [ ] Ambient Weather Station - Android _(Add in My Device Handlers)_

2. Change the device TYPE used for your Ambient Weather Station device to 
- [ ]  Ambient Weather Station - Android _(Update in My Devices_)

![image](https://user-images.githubusercontent.com/12133562/49284604-0d7c7500-f463-11e8-990f-4d7578b26df2.png)

**Comment by jeffp1105** [comment]
HI Kurt,

Here's what I found so far.  Forgive me for not being ST developer and some
possible naive on the platform.

The install issues are fixed with the patches you made.  For thoroughness,
I removed the older app and DTH. As a result, I could not re-install the
new app because it first looks for the old dth to install which is now
gone. Simple enough, I changed my version of the code to default to the new
driver.

There's a minor issue: every once in a while I see an uncaught null pointer
for the moon data.  It's rare enough that I am ignoring it for now.

The new app and DHT load and to confirm the new DHT is used I added my
initials to some log messages. However, the result is the same.  Next, I
added a whole bunch of debug statements (log.info) all over the DHT. What I
found was the only 2 messages ever received are "DTH Section installed" and
DTH section unistalled". When I tap the display in the classic app, that's
when the error screen appears in the app (but no messages in live logging)
and when I tap the tile in the new ST App, literally nothing happens.
There is no sign in the logs that the DTH initialize, refresh, or updated
are ever run.

Let me know if you need screens, logs, or other tests.

*Jeff*

**Comment by KurtSanders** [comment]
Dear Jeff,

Sorry for the confusion and resultant work on your part to accomodate this debugging.   

I did not want you to have to remove and reinstall the SmartApp, but simply just change the **device name** used for the 'Ambient Weather Station device' in your IDE 'My Devices' to **Ambient Weather Station - Android**  Each device listed in your My Devices view can be edited and the device name for the DTH can be changed via a drop down box.    Hopefully your Ambient Weather Station record looks like this below which is using the **Ambient Weather Station - Android**

![image](https://user-images.githubusercontent.com/12133562/49322269-2a08c380-f4dc-11e8-81c9-bffa23d9329e.png)

So, I have made some more changes to the **Ambient Weather Station - Android** DTH, and when you update the DTH from the Repo  for the "My Device Handlers", let me know if you can display the full tile.  Here is the way to update the **Ambient Weather Station - Android** DTH.  Make sure you check the **Publish** check box.

![image](https://user-images.githubusercontent.com/12133562/49322335-9be10d00-f4dc-11e8-846c-1f5e66e58e35.png)
![image](https://user-images.githubusercontent.com/12133562/49322414-54a74c00-f4dd-11e8-93ec-22feff37945c.png)

Thanks again, I wish I had a way to test this on an Android device so to get you a display that works.  It's maddening that it works on the iPhone client.

After we get the **Ambient Weather Station - Android** device Handler (DTH) to display on Android devices, I will update the SmartApp to point the correct name for the final working DTH.  
### Fingers crossed ;)

**Comment by KurtSanders** [comment]
Dear Jeff,

I will be out of town for next 5 days, looking forward to your test results when I return to close this nagging issue.

**Comment by jeffp1105** [comment]
Hi Kurt,

Sorry to go dark. I've been busy focusing on some other home automation
projects (aka, honey-dew list).I had worked on this over the weekend
without much success. I didn't want to give you half-baked info so didn't
give a status.  I will have time tomorrow and plan to start from scratch.
What I saw so far was that whether  I hit the tile in the new app or hit
the "thing" in the old app it appears none of the code in the DHT runs.  I
need to figure out a way to make sure the updated DHT is installed
correctly.  Anyway, will get back to you with some pics and better detail
by your return.

Thanks again,
Jeff

*Jeff Posner*
*2938 Casalon Circle*
*Superior, CO 80027*
Home: 703-929-3736
Work: 202-794-4366

**Comment by KurtSanders** [comment]
This is crazy, I’m not sure why this DTH won’t load in the Android client, very frustrating for both of us…. Well, I might submit it to the ST Forum developers and see if they can spot the place in the DTH that is problematic to Android ST Client.  Thanks for testing and I can keep you in the loop if I hear anything that changes where we are.

When you check your My Devices tab for the Ambient Weather Station, does it show the Android version of the DTH?

**Comment by jeffp1105** [comment]
Yes, for sure it says - Android.  TO be clear, I added log.info statements
and my initials to the set of functions at the bottom of the driver code.
What I see is that the "Installed" method is called and if I remove the
Smartapp, the DTH is removed and I do so the "uninstalled" message.  I
never see the updated, initialize or refresh methods.  So I would focus on
what triggers the Android app to actually execute these other methods.  I'm
just guessing here. I suspect there's some Android trick or technique
that's missing.

[image: Screen Shot 2018-12-04 at 9.45.02 AM.jpg]

*Jeff Posner*
*2938 Casalon Circle*
*Superior, CO 80027*
Home: 703-929-3736
Work: 202-794-4366

**Comment by KurtSanders** [comment]
The ambient Android DTH tile does not have any active code and does not execute those required sections.  It does have some standard statements for tile color based on weather values.  All of the weather values come from the Ambient Weather Station Service Manager which connects to your Ambient personal weather station cloud server and obtains the local weather.  

That is why I am so confused as to why Android devices cannot simply display the tile like the iOS devices as they do with so many other DTH's.  Not having a Android device is hard to debug any further so I will ask the developers forum for their thoughts, as I really would like to crack this bug and simply have ONE DTH for all devices...  ST engineering should be ashamed that the Android version of the ST mobile client is so disappointing...

**Comment by KurtSanders** [comment]
Posted this issue on SmartThings forum

https://community.smartthings.com/t/please-help-ambient-weather-device-handler-dth-works-for-apple-ios-st-client-but-not-for-android-o-s/145062

**Comment by KurtSanders** [comment]
Dear Jeff,

While we are waiting on ST developers to test and hopefully identify the issue with this Ambient Weather Station tile on Android devices.  Have you ever tried installing the "Weather Station Tile 2.0 App" to see if it displays accurately on your Android?

https://community.smartthings.com/t/release-weather-station-tile-2-0/43833

**Comment by jeffp1105** [comment]
Yes, this one runs.  I believe it pulls data from Weather Underground
rather than from the Ambient Servers.  I'd rather use your tile since it
pulls directly from my (our) weather vendor instead of relying on an
intermediate server.

J

PS: If you are interested, I think I have a Galaxy S5 and I know I have an
S7 lying around that I don't use.  I was planning to set them up as "remote
controls" running Action Tiles but I've been informed by my kids they don't
want a "smart house".  Both work fine, no cracks.  I could sell one cheap
if you want to have an android as a development platform.

*Jeff Posner*
*2938 Casalon Circle*
*Superior, CO 80027*
Home: 703-929-3736
Work: 202-794-4366

**Comment by KurtSanders** [comment]
Dear Jeff,

Good to know that that Weather Underground runs on Android, so I can see what that DTH has for some settings.  

I made **One More Change** to the **Ambient Weather Station Android DTH** and **Ambient Weather Station Service Manager smartApp** and both DTH and SmartApp are NOW available via the "Update from Repo" function in the ST IDE.  If you want to upgrade to the latest beta version(s), it might be one possible working model for your beta testing.  A few ST developers have been assisting me with some areas that can be problematic for Android clients.  Let me know if this new code changes to shwoing you the full tiles!

I might be interested in the one of your units, name a fair price and let's see if we can work out a deal!

**Comment by KurtSanders** [comment]
Looks like we made some progress finding the specific issue, and the latest version of the Android DTH and Service Manager might give you a complete weather station view!  This developer Orangebucket was helpful in pointing me to the problematic Android statements.   Let me know if the newest version of the DTH and Service Manager SmartApp available in the "Update from Repo" works!

![image](https://user-images.githubusercontent.com/12133562/49610387-6ae55a00-f96c-11e8-8831-fece60a77353.png)

**Comment by jeffp1105** [comment]
As Gene Wilder said in in "Young Frankenstein". IT IS ALIVE!

[image: Screenshot_20181207-201732_SmartThings.jpg]
[image: Screenshot_20181207-201741_SmartThings.jpg]

*Jeff Posner*
*2938 Casalon Circle*
*Superior, CO 80027*
Home: 703-929-3736
Work: 202-794-4366

**Comment by KurtSanders** [comment]
### Wonderful...  So glad!   

In a few days, I will merge the 'Master' and 'Beta" branches of the Ambient Weather Station Github repositories to only have a V1 (no service Manager) and V2 (Service Manager Req) DTH's.  You can stay with what is working for your Android and not change, or follow the new install/update instructions to keep current.  I'm not sure I will add more significant functions to V2, but I have been tweaking some weather station variables and tile customizations over the last month.

**Again, I might be interested in the one of your Android phone units, if you name a fair price, we can work out a deal!**

**Comment by jeffp1105** [comment]
Can we private message about the phones. You can see my email in my profile?

Do you plan to create a combined tile like the ST "weather" tile? The one
tile has about six values in one and is useful in Actiontiles.  Somehow the
ST advertises an additional tile to AT called "weather". Sorry I'm not up
on the terminology but can send pics later today and you'd probably
recognize what's happening.

**Comment by jeffp1105** [comment]
Can we private message about the phones. You can see my email in my profile?

Do you plan to create a combined tile like the ST "weather" tile? The one
tile has about six values in one and is useful in Actiontiles.  Somehow the
ST advertises an additional tile to AT called "weather". Sorry I'm not up
on the terminology but can send pics later today and you'd probably
recognize what's happening.

On Dec 8, 2018 4:20 AM, "Kurt Sanders" <notifications@github.com> wrote:

Wonderful... So glad!

In a few days, I will merge the 'Master' and 'Beta" branches of the Ambient
Weather Station Github repositories to only have a V1 (no service Manager)
and V2 (Service Manager Req) DTH's. You can stay with what is working for
your Android and not change, or follow the new install/update instructions
to keep current. I'm not sure I will add more significant functions to V2,
but I have been tweaking some weather station variables and tile
customizations over the last month.

*Again, I might be interested in the one of your Android phone units, if
you name a fair price, we can work out a deal!*

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/KurtSanders/STAmbientWeather/issues/3#issuecomment-445451824>,
or mute the thread

**Comment by KurtSanders** [comment]
Sounds good to me, but I do not see your email or a means to PM on Github.  Am I missing something?

**Comment by jeffp1105** [comment]
My email is jeffp1105 at that email run by Google
:)

**Comment by KurtSanders** [comment]
The beta and master Ambient Weather Station branches have been merged into the 'master' branch.  It is strongly recommended that those ST users who were testing this application in the 'beta' branch return their ST IDE's 'Update from Repo' repository settings to 'master' & synchronize both the DTH and SmartApp with the 'master' branch.  

The 'beta' branch has been deleted and all future updates to **STAmbientWeather** will occur in the 'master' branch.

| Name | Value |
|------|-------|
|Owner | **kurtsanders** |
|Name: | **STAmbientWeather**|
|Branch| **master**|
